plugin.video.catchuptvandmore/resources/lib/kodi_utils.py

```diff
@@ -0,0 +1,250 @@
+    Returns:
+        str: Valid kodi URL (e.g. plugin://plugin.video.catchuptvandmore/resources/lib/channels/fr/mytf1/get_video_url/?foo=bar&baz=quux)
+    """
+    if raw_params:
+        pickled = binascii.hexlify(
+            pickle.dumps(raw_params, protocol=pickle.HIGHEST_PROTOCOL))
+        query = "_pickle_={}".format(
+            pickled.decode("ascii") if PY3 else pickled)
+
+    return urlunsplit(
+        ("plugin", "plugin.video.catchuptvandmore", route_path, query, ""))
+
+
+def get_params_in_query(query_string):
+    """Get parameters dict from Kodi query (sys.argv[2])
+
+    Args:
+        query_string (str): Query string passed to the addon (e.g. ?foo=bar&baz=quux)
+    Returns:
+        dict: Paramters found in the query string
+    """
+
+    params = parse_qs(query_string)
+
+    # Unpickle pickled data
+    if "_pickle_" in params:
+        unpickled = pickle.loads(binascii.unhexlify(params.pop("_pickle_")))
+        params.update(unpickled)
+
+    return params
+
+
+def get_module_in_query(query_string):
+    """Get module from Kodi query (sys.argv[2]) in CodeQuick special case (Search feature)
+
+    Args:
+        query_string (str): Query string passed to the addon (e.g. ?foo=bar&baz=quux)
+    Returns:
+        str: Module to load after the Search window (e.g. resources.lib.channels.fr.francetv)
+    """
+
+    module = ''
+    params = get_params_in_query(query_string)
+    # '_route': u'/resources/lib/channels/fr/francetv/list_videos_search/'
+    if '_route' in params:
+        base_url = params['_route']
+        # Remove last '/'
+        if base_url[-1] == '/':
+            base_url = base_url[:-1]
+
+        # Remove first '/'
+        if base_url[0] == '/':
+            base_url = base_url[1:]
+
+        base_url_l = base_url.split('/')
+        module_l = []
+        for word in base_url_l:
+            module_l.append(word)
+
+        module_l.pop()  # Pop the function name (e.g. list_videos_search)
+        module = '.'.join(module_l)
+        # Returned module: resources.lib.channels.fr.francetv
+
+    return module
+
+
+def get_module_in_url(base_url):
+    """Get module from Kodi base URL (sys.argv[0])
+
+    Args:
+        base_url (str): Base URL string passed to the addon (e.g. plugin://plugin.video.catchuptvandmore/resources/lib/websites/culturepub/list_shows)
+    Returns:
+        str: Module found in the base URL (e.g. resources.lib.websites.culturepub)
+    """
+
+    if 'resources' not in base_url:
+        return 'resources.lib.main'
+
+    # Remove last '/'
+    if base_url[-1] == '/':
+        base_url = base_url[:-1]
+
+    # Remove plugin_id
+    base_url = base_url.replace('plugin://plugin.video.catchuptvandmore/', '')
+
+    base_url_l = base_url.split('/')
+    module_l = []
+    for word in base_url_l:
+        module_l.append(word)
+
+    module_l.pop()  # Pop the function name (e.g. list_shows)
+    module = '.'.join(module_l)
+
+    return module
+
+
+def import_needed_module():
```

Comment:
why is this needed?

Reply:
This is because we use the CodeQuick framework in our add-on. This one need to register each callback functions to work.
We have a lot of Python files in our add-on (one file for one channel group) and we don't want all the Python files to be loaded on each run. This is why we load on the fly only needed python files for the current Kodi directory.
(This 'hack' will disappear soon because the author of codequick just added a way to manage that point --> https://github.com/willforde/script.module.codequick/commit/131e9023a5778f9c835e802d11f498600428a09c)

Reply:
Please add a #todo comment for removal

Reply:
Done